### ISSUE TITLE
router: protect `routeall()` and others from modification

### ISSUE THREAD

**Issue by Serpentian** [tarantool/vshard#502]
The function `vshard.router.routeall` is not marked as internal API (like it's done for `vshard.router.bucket_discovery` e.g.), but it returns `router.replicasets` table, which is used by the router itlself. If user modifies the returned from `routeall` table (e.g. deletes some replicaset just because he wants to skip it during request), then router is broken and cannot continue to work properly after that. Moreover, these types of errors are hard to find.

The same applies to `vshard.router.route` and `vshard.router.bucket_discovery` (internal one) functions.

### Proposed solution

Create a special table of replicasets during router configuration, so that not to recreate it every time when we need to return replicasets to user. This table doesn't allow inserting of a new value or deleting of the existing one (`__newindex`), as otherwise the change of that table will be preserved between `routeall`.

Every replicaset is wrapped (it may be returned as value in `routeall` table or from `route`): replicaset has only one real field: `_replicaset`, which is the one we return now. It has all of the supported replicaset functions in the metatable. However, user usually needs to access fields of the replicaset (e.g. `bucket_count`) and we don't want user to end up accessing `_replicaset` field. For that we modify the `__index` of the replicaset wrapper, allowing to access safe fields: all fields, except `master_cond`, `on_master_required` (they can be overwritten). It's forbidden to add new fields (or change `_replicaset`) to replicaset wrapper.

In that wrapper user still needs to access replica objects. So, it must also be wrapped. Still, all functions of the replica's metatable; only `_replica` real field, all internal fields can be accessed via `__index`, except `conn`. Link in replica `next_by_priority` points to the wrapper. `master` and `replica` in replicaset wrapper point to replica wrappers. `priority_list` and `replicas` in replicaset wrapper are the tables of wrappers, it's forbidden to modify them. It's forbidden to add new fields to replica wrapper.

This way user won't be able to modify the fields of returned tables (if he won't access `rs._replicaset` or `r._replica`, which he should not if he doesn't understand, what he's doing). But the table is RO, which may not be very convenient (see [alternative](#modify)).

### Considered alternatives

1. `table.deepcopy`

`table.copy` or `table.deepcopy` may seem like a simple solution, but it's not. Connections cannot be copied. User will have to reconnect for every request after every `routeall` in such case.

2. Allow to modify returning tables <a name=#modify></a>

The alternative solution is to create the wrapper on every `routeall` or `route` call and to wrap the replicaset on every `route` call. This way, user will be able to modify the returned by `routeall` or `route` table as he likes, we won't forbid adding new fields or deleting the old ones. It makes user experience less painful but requires creating it on every call (is it that costly? @Gerold103).

**Comment by Serpentian** [comment]
Related: https://github.com/tarantool/vshard/issues/503

**Comment by Gerold103** [comment]
> In that wrapper user still needs to access replica objects.

Nope. User shouldn't access individual replica objects ever and it isn't public. At least it was never intended. I suggest to forbid that for now completely. The replicaset objects can only do `call/callro/callrw`. No members of them are public. Only those functions.

> is it that costly?

It is expensive, I think. If replicaset count is big enough - 5, 10 is ok, but is 100 ok? Besides, it isn't only about costs. The function says `routeall()`. The necessity to delete or add something already means the user doesn't need "all". For such guys we could have another method if this feature is in high demand. And to wrap it up - where would this editing end? Do we allow to remove the replicasets from the root table, but protect each replica object in it? Do we allow to modify the connections, or modify the replicaset objects? The simplest way is to just forbid it all. Those who try would get an error, and would make their own table with UUIDs for example. We do that in vshard as well in many places.

We can think of a more reusable way if this becomes frequently asked for, ok?

**Comment by Serpentian** [comment]
Related https://github.com/tarantool/vshard/issues/494

**Comment by Serpentian** [comment]
This one is a breaking change, crud will stop working after new `routeall` is merged. It widely uses `routeall`
